src/lib.rs

```diff
@@ -3233,6 +3233,11 @@ impl Connection {
             .is_none()
     }
 
+    /// Returns whether there is stream or DATAGRAM data available to read.
+    pub fn is_app_data_readable(&self) -> bool {
+        self.readable().next().is_some() || self.dgram_recv_front_len().is_some()
```

Comment:
You can just do `self.readable().len() > 0` which is slightly simpler, but more generally creating the iterator is not free, and then it gets discarded so we are basically paying the price for nothing here.

A better alternative would be to add a `has_readable()` method like https://github.com/cloudflare/quiche/blob/master/src/stream.rs#L431 and then call that from here, so this method becomes effectively free.

Reply:
You can't get length of the iterator without collecting it. Also, `readable` doesn't give information about pending dgrams

Reply:
But I agree that it would be even better without dealing with the iterator if that's possible.

Reply:
Just FTR, `StreamIter` implements `ExactSizeIterator` https://doc.rust-lang.org/nightly/core/iter/trait.ExactSizeIterator.html so you can get the length without collecting I think.

Reply:
(also, I meant use `len()` for the `self.readable()` part, the dgram check would still be needed).